### PR TITLE
Refactor TestE2E to work correctly in parallel mode

### LINKED ISSUE #13485: E2e: jenkins test run timeouts without clear reason
Both the GKE and Parallel tests recently had a run where they timed out. (after 90 minutes!) This is like 10x longer than the parallel run should be taking.

http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gke-ci/8168/

http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce-parallel/7141/

### PR DESCRIPTION
**Pull request by ixdy**
The TestE2E function is run on every Ginkgo parallel node.
The cluster cleanup, sanity checking, and log dump operations are only
intended to be called once, however, so these have been moved out into
SynchronizedBeforeSuite and SynchronizedAfterSuite blocks.

I'm hoping this fixes #13485.

Depending on who wins the merge race, either this PR or #19624 will get to have a fun rebase.

cc @kubernetes/goog-testing @jayunit100

### CODE REVIEW

**Comment by ixdy** [comment]
Would be good to get this in soon if possible (sorry @jayunit100), since it seems that Ginkgo is mysteriously timing out more and more, and I'd like to know if this fixes or if there's something else wrong.

**Comment by jayunit100** [comment]
thanks for the heads up. If you beat me to it do you mind squeezing a review in this wknd so I can rebase? Our libification patch is super critical also .

Cc @timothysc

**Comment by ixdy** [comment]
rebased

File: test/e2e/e2e_test.go [hunk 1]
```diff
@@ -180,14 +169,55 @@ func TestE2E(t *testing.T) {
...
+	// Run on all nodes
+
+}, func() {
+	// Run only on node 1
+	if testContext.ReportDir != "" {
+		CoreDump(testContext.ReportDir)
+	}
+})
+
+func TestE2E(t *testing.T) {
```
Comment (ihmccreery) [inline test/e2e/e2e_test.go]: Public function should have a comment explaining what it does.

File: test/e2e/e2e_test.go [hunk 2]
```diff
@@ -121,57 +112,55 @@ func TestE2E(t *testing.T) {
...
 
-	c, err := loadClient()
-	if err != nil {
-		glog.Fatal("Error loading client: ", err)
-	}
+	return nil
+}
+
+var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
+	// Run only on node 1
```
Comment (ihmccreery) [inline test/e2e/e2e_test.go]: Why?

File: test/e2e/e2e_test.go [hunk 3]
```diff
@@ -121,57 +112,55 @@ func TestE2E(t *testing.T) {
...
-	gomega.RegisterFailHandler(ginkgo.Fail)
 
-	c, err := loadClient()
-	if err != nil {
-		glog.Fatal("Error loading client: ", err)
-	}
+	return nil
+}
+
+var _ = ginkgo.SynchronizedBeforeSuite(func() []byte {
```
Comment (ihmccreery) [inline test/e2e/e2e_test.go]: This is a wonky enough syntax and procedure that I think it deserves a comment.  What are you doing here & why?

**Comment by ihmccreery** [comment]
LGTM, but I'm having a hard time following.  Please add comments in code, not just in PR.

**Comment by ixdy** [comment]
Thanks for the review! I've added comments. Let me know if you are happy with them and I'll squash.

File: test/e2e/e2e_test.go [hunk 4]
```diff
@@ -180,14 +177,63 @@ func TestE2E(t *testing.T) {
...
+		// TODO: we should probably only be trying to create this directory once
+		// rather than once-per-node.
+		if err := os.MkdirAll(testContext.ReportDir, 0755); err != nil {
+			glog.Errorf("Failed creating report directory: %v", err)
+		} else {
+			r = append(r, reporters.NewJUnitReporter(path.Join(testContext.ReportDir, fmt.Sprintf("junit_%02d.xml", config.GinkgoConfig.ParallelNode))))
+		}
 	}
-	glog.Infof("Starting e2e run; %q", runId)
+	glog.Infof("Starting e2e run %q on node %d", runId, config.GinkgoConfig.ParallelNode)
```
Comment (ihmccreery) [inline test/e2e/e2e_test.go]: "node" is overloaded in this context; maybe specify "Ginkgo node"?

**Comment by ihmccreery** [comment]
LGTM with minor nit; squash and feel free to apply `lgtm`.

**Comment by ixdy** [comment]
Squashed and added "Ginkgo " in front of "node" in the `Infof` as well as comments.

Applying lgtm label per comment.

**Comment by ixdy** [comment]
Last failure was flake #18210.

**Comment by ixdy** [comment]
@k8s-oncall this should be safe to merge, and hopefully fixes some e2e timeouts (like kubernetes-pull-build-test-e2e-gce/25483 and kubernetes-pull-build-test-e2e-gce/25484 yesterday).

**Comment by fabioy** [comment]
@ixdy I'll merge this soon. Will this affect #19970?

**Comment by fabioy** [comment]
@ixdy, sorry, meant #19624.

**Comment by ixdy** [comment]
Yes, whichever one of these merges first will have to rebase on the other.

**Comment by ixdy** [comment]
Er. Whichever merges first will force the other one to rebase.

**Comment by fabioy** [comment]
Ok, merging this one first. The other one already needs a rebase...

**Pull request merged**